### ISSUE TITLE
Mypy points missing return statement when explicity annotated with `A | None`

### ISSUE THREAD

**Issue by betafcc** [python/mypy#13855]
**Bug Report**

Noticed some async gen warnings inconsistency between pyright and mypy, mypy complains about a `return` inside async for loop with no fallback after it:

**To Reproduce**
Mypy complains about

```python
from typing import AsyncIterable, TypeVar

A = TypeVar("A")

async def first(fa: AsyncIterable[A]) -> A | None:
    async for a in fa:
        return a
```

Do note the explicit `A | None` annotation, I do recognize the bad practice here but in typing it seems that shouldn't be wrong, pyright passes as expected but mypy gives:

```sd
❱ mypy dirty_bits/helpers/aitertools.py 
dirty_bits/helpers/aitertools.py:127: error: Missing return statement
```

**Your Environment**

- Mypy version used: mypy 0.982 (compiled: yes)
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10.6

related: https://github.com/microsoft/pyright/issues/4027

**Comment by Avasam** [comment]
Are you looking for the `--no-warn-no-return` flag?

```
Configuring warnings:
  Detect code that is sound but redundant or problematic.

  --no-warn-no-return       Do not warn about functions that end without returning (inverse: --warn-no-return)
```

**Comment by hauntsaninja** [comment]
Avasam is correct

**Comment by betafcc** [comment]
> Are you looking for the `--no-warn-no-return` flag?
> 
> ```
> Configuring warnings:
>   Detect code that is sound but redundant or problematic.
> 
>   --no-warn-no-return       Do not warn about functions that end without returning (inverse: --warn-no-return)
> ```

oh sorry, didn't knew about that
